Support multiple mappers for each class

iruizmar opened the pull request:
Closes #70

luca992 reviewed (changes requested):
Looks good besides that

luca992 commented on kamel-core/src/commonTest/kotlin/io/kamel/core/config/KamelConfigBuilderTest.kt:
You should just be able to use flatten

iruizmar replied:
@luca992 I can change it, but this way it's also testing that the mappers are added to the correct key. Is it over-testing in your opinion?

luca992 replied:
hmm true. technically that's tested with lines above in the single mapper tests at least. But I guess its good to be more explicit. could you just use getOrDefault to return an empty list to remove the assert not null then?

iruizmar replied:
@luca992 I tried your syntax and found a less verbose one, LMK if you like it.

luca992 replied:
Looks good

luca992 commented on kamel-core/src/commonMain/kotlin/io/kamel/core/config/KamelConfigBuilder.kt:
```diff
@@ -48,7 +48,8 @@ public class KamelConfigBuilder {
     public fun <I : Any, O : Any> mapper(mapper: Mapper<I, O>) {
-        mappers[mapper.inputKClass] = mapper as Mapper<Any, Any>
+        mappers[mapper.inputKClass] =
+            (mappers[mapper.inputKClass] ?: emptyList()) + (mapper as Mapper<Any, Any>)
```
actually you can use `getOrPut` here too

iruizmar replied:
Had to make the list of mappers a `MutableList`, though.

luca992 commented on kamel-core/src/commonMain/kotlin/io/kamel/core/config/KamelConfigBuilder.kt:
```diff
@@ -128,7 +130,7 @@ public fun KamelConfigBuilder.takeFrom(config: KamelConfig): KamelConfigBuilder
     config.fetchers.forEach { fetcher(it) }
     config.decoders.forEach { decoder(it) }
-    config.mappers.values.forEach { mapper(it) }
+    config.mappers.values.forEach { mappers -> mappers.forEach { mapper(it) } }
```
And use flatten here

luca992 commented on kamel-core/src/commonMain/kotlin/io/kamel/core/utils/ConfigUtils.kt:
```diff
@@ -8,7 +8,11 @@ import io.kamel.core.fetcher.Fetcher
-    val output = mappers[inputKClass]?.map(input)
+
+    val output = mappers[inputKClass]
+        ?.firstOrNull { mapper -> with(mapper) { input.isSupported } }
```
and sorry, use `lastOrNull` to match the behavior of fetchers.

luca992 commented:
Sorry, I looked at it again, and made more change requests

iruizmar commented:
@luca992 Sure! I think you forgot to send them, though.

luca992 commented:
Alright, once again, sorry I'm thinking there might be a better way instead of using:

`val mappers: MutableMap<KClass<*>, List<Mapper<Any, Any>>> = mutableMapOf()`

to use:

`val mappers: MutableMap<Pair<KClass<*>, KClass<*>>, Mapper<Any, Any>> = mutableMapOf()`

what do you think?

luca992 commented:
> > Sorry, I looked at it again, and made more change requests
> 
> @luca992 Sure! I think you forgot to send them, though.

I think you should be able to see them above. Lmk if you do not

iruizmar commented:
> > > Sorry, I looked at it again, and made more change requests
> > 
> > 
> > @luca992 Sure! I think you forgot to send them, though.
> 
> I think you should be able to see them above. Lmk if you do not

Can't see anything, mate! Maybe you started a review and didn't send it?

luca992 commented:
> > > > Sorry, I looked at it again, and made more change requests
> > > 
> > > 
> > > @luca992 Sure! I think you forgot to send them, though.
> > 
> > 
> > I think you should be able to see them above. Lmk if you do not
> 
> Can't see anything, mate! Maybe you started a review and didn't send it?

Sorry, thought I added them to the old review. I guess not.

iruizmar commented:
> > > > > Sorry, I looked at it again, and made more change requests
> > > > 
> > > > 
> > > > @luca992 Sure! I think you forgot to send them, though.
> > > 
> > > 
> > > I think you should be able to see them above. Lmk if you do not
> > 
> > 
> > Can't see anything, mate! Maybe you started a review and didn't send it?
> 
> Sorry, thought I added them to the old review. I guess not.

@luca992 all good. I addressed them.

luca992 commented:
Sweet looks good. Thanks!
